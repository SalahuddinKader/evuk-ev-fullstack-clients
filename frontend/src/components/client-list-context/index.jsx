import React, { useState, useEffect, createContext } from "react";
import api from "../../api";
/**
 * Client list context display all state and functions
 * @state {array} clients
 * @state {string} search
 * @state {array} searchResult
 * @func {function} AddClientHandler
 * @func {function} retrieveClients
 * @func {function} removeClientHandler
 * @func {function} searchHandler
 */
export const ClientContext = createContext();

export const ClientListContext = (props) => {
  const [clients, setClients] = useState([]);
  const [search, setSearch] = useState("");
  const [searchResult, setSearchResult] = useState([]);

  // AddClientHandler and post clients
  const addClientHandler = async (client) => {
    const request = {
      id: Math.random() * 1000,
      ...client,
    };
    const response = await api.post("/clients", request);
    setClients([...clients, response.data]);
  };

  // Handle get api
  const retrieveClients = async () => {
    const response = await api.get("/clients");
    return response.data;
  };

  useEffect(() => {
    const getAllClients = async () => {
      const allClients = await retrieveClients();
      if (allClients) setClients(allClients);
    };
    getAllClients();
  }, []);

  // Delete individual client
  const removeClientHandler = async (id) => {
    await api.delete(`/clients/${id}`);
    setClients(clients.filter((client) => client.id !== id));
  };
  // Handling search through object value
  const searchHandler = (search) => {
    setSearch(search);
    if (search !== "") {
      const newClientList = clients.filter((client) => {
        return Object.values(client)
          .join(" ")
          .toLowerCase()
          .includes(search.toLowerCase());
      });
      setSearchResult(newClientList);
    } else {
      setSearchResult(clients);
    }
  };

  return (
    <>
      <ClientContext.Provider
        data-testid="client-list-context-provider"
        value={{
          clients,
          search,
          searchResult,
          searchHandler,
          setClients,
          addClientHandler,
          retrieveClients,
          removeClientHandler,
        }}
      >
        {props.children}
      </ClientContext.Provider>
    </>
  );
};

export default ClientListContext;
