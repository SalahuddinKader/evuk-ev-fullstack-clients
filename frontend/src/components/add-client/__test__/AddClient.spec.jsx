import React from 'react';
import { render } from '@testing-library/react';
import { ClientContext } from '../../client-list-context';
import { BrowserRouter } from 'react-router-dom';
import AddClient from '..';

describe('AddClient Component', () => {
  it('matches the snapshot', () => {
    const { baseElement } = render(
      <ClientContext.Provider
        value={{
          clients: [
            {
              id: 527.2950507590961,
              name: 'ClientName',
              email: 'client@example.com',
              companyName: 'ABC',
              streetAddress: '20 Courtney Road',
              postCode: 'RM164TY',
              city: 'City',
              phone: '0123456789',
            },
          ],
          search: '',
          searchResult: [],
          searchHandler: () => {},
          setClients: () => {},
          AddClientHandler: () => {},
          retrieveClients: () => {},
          removeClientHandler: () => {},
        }}
      >
        <BrowserRouter>
          <AddClient />
        </BrowserRouter>
      </ClientContext.Provider>
    );
    expect(baseElement).toMatchSnapshot();
  });
});
