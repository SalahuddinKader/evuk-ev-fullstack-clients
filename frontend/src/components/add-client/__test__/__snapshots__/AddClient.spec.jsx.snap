// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddClient Component matches the snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
      data-testid="card-add-client"
      style="max-width: 500px; margin: 0px auto; padding: 20px 5px;"
    >
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          data-testid="add-client-label"
        >
          Add Client
        </h4>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-ev0a51-MuiGrid-root"
        >
          <a
            href="/"
            style="text-decoration: none;"
          >
            <button
              class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
              data-testid="button-back-to-list"
              tabindex="0"
              type="button"
            >
              Back to list
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
        </div>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-186q1z4-MuiTypography-root"
        >
          Fill the form to Add new Client
        </p>
        <form
          data-testid="form-submit"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-a5rdam-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                data-testid="name-text-field"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                >
                  Name
                  <span
                    aria-hidden="true"
                    class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    name="name"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-173wfxe"
                    >
                      <span>
                        Name
                         *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                data-testid="email-text-field"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                >
                  Email
                  <span
                    aria-hidden="true"
                    class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    name="email"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-173wfxe"
                    >
                      <span>
                        Email
                         *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                data-testid="company-name-text-field"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                >
                  Company Name
                  <span
                    aria-hidden="true"
                    class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    name="companyName"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-173wfxe"
                    >
                      <span>
                        Company Name
                         *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                data-testid="phone-text-field"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                >
                  Phone
                  <span
                    aria-hidden="true"
                    class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    name="phone"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-173wfxe"
                    >
                      <span>
                        Phone
                         *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                data-testid="street-text-field"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                >
                  Street
                  <span
                    aria-hidden="true"
                    class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    name="streetAddress"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-173wfxe"
                    >
                      <span>
                        Street
                         *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                data-testid="postcode-text-field"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                >
                  Postcode
                  <span
                    aria-hidden="true"
                    class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    name="postCode"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-173wfxe"
                    >
                      <span>
                        Postcode
                         *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                data-testid="city-text-field"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                >
                  City
                  <span
                    aria-hidden="true"
                    class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    name="city"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-173wfxe"
                    >
                      <span>
                        City
                         *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-5.1 css-191kdtn-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                >
                  Date
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    aria-label="Choose date"
                    aria-readonly="true"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-173wfxe"
                    >
                      <span>
                        Date
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-uvohki-MuiGrid-root"
            >
              <button
                class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
                data-testid="button-add"
                tabindex="0"
                type="submit"
              >
                Add
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</body>
`;
